src/app/consensus_test/consensus_test.ml

```diff
@@ -0,0 +1,248 @@
+  in
+  let () =
+    if not (Core.Sys.file_exists_exn output_dir) then Core.Unix.mkdir output_dir
+  in
+  Deferred.List.iteri sorted_output ~f:write_block_to_file
+
+let compare_lengths candidate_length existing_length =
+  if candidate_length > existing_length then Candidate_longer
+  else if candidate_length = existing_length then Equal_length
+  else Candidate_shorter
+
+let run_select ~context (existing_block : Mina_block.Precomputed.t)
+    (candidate_block : Mina_block.Precomputed.t) =
+  let existing_consensus_state_with_hashes =
+    { With_hash.hash =
+        Mina_state.Protocol_state.hashes existing_block.protocol_state
+    ; data =
+        Mina_state.Protocol_state.consensus_state existing_block.protocol_state
+    }
+  in
+  let candidate_consensus_state_with_hashes =
+    { With_hash.hash =
+        Mina_state.Protocol_state.hashes candidate_block.protocol_state
+    ; data =
+        Mina_state.Protocol_state.consensus_state candidate_block.protocol_state
+    }
+  in
+  match
+    Consensus.Hooks.select ~context
+      ~existing:existing_consensus_state_with_hashes
+      ~candidate:candidate_consensus_state_with_hashes
+  with
+  | `Take ->
+      let candidate_length =
+        Mina_state.Protocol_state.consensus_state candidate_block.protocol_state
+        |> Consensus.Data.Consensus_state.blockchain_length
+        |> Unsigned.UInt32.to_int
+      in
+      let existing_length =
+        Mina_state.Protocol_state.consensus_state existing_block.protocol_state
+        |> Consensus.Data.Consensus_state.blockchain_length
+        |> Unsigned.UInt32.to_int
+      in
+      compare_lengths candidate_length existing_length
+  | `Keep ->
+      Candidate_shorter
+
+let update_chain ~current_chain ~candidate_block ~select_outcome =
+  match select_outcome with
+  | Candidate_longer ->
+      candidate_block :: current_chain
+  | Equal_length -> (
+      match current_chain with
+      | _ :: rest_of_list ->
+          candidate_block :: rest_of_list
+      | [] ->
+          current_chain )
+  | Candidate_shorter ->
+      current_chain
+
+let process_precomputed_blocks ~context ~current_chain blocks =
+  List.fold blocks ~init:current_chain ~f:(fun acc_chain candidate_block ->
+      let existing_block = List.hd_exn acc_chain in
+      let select_outcome = run_select ~context existing_block candidate_block in
+      update_chain ~current_chain:acc_chain ~candidate_block ~select_outcome )
+
+let extract_timestamp (block : Mina_block.Precomputed.t) =
+  let bs = Mina_state.Protocol_state.blockchain_state block.protocol_state in
+  Block_time.to_time_exn bs.timestamp
+
+let main () ~blocks_dir ~output_dir ~runtime_config_file =
+  let logger = Logger.create () in
+  let%bind context = generate_context ~logger ~runtime_config_file in
+  let%bind block_sorted_filenames = read_directory ~logger blocks_dir in
+  let precomputed_blocks =
+    block_sorted_filenames
+    |> List.map ~f:read_block_file
+    |> List.filter_map ~f:Fun.id
```

Comment:
think you can do one iteration over the list, instead of filtering in a separate step